function() {
    try {
    } finally {
        return;
        ~~~~~~~    [no-return-in-finally]
    }
}

function() {
    try {
    } catch {
    } finally {
        return 1;
        ~~~~~~~~~    [no-return-in-finally]
    }
}

function() {
    try {
    } catch {
    } finally {
        {
            return 1;
            ~~~~~~~~~    [no-return-in-finally]
        }
    }
}

function() {
    try {
    } catch {
    } finally {
        for (let i = 0; i < 5; ++i) {
            if (i % 2 === 0) {
                return 1;
                ~~~~~~~~~    [no-return-in-finally]
            }
        }
    }
}

function() {
    try {
    } catch {
    } finally {
        let i = 1;
        while (i < 5) {
            if (i % 2 === 0) {
                return i;
                ~~~~~~~~~    [no-return-in-finally]
            }

            ++i;
        }
    }
}

function() {
    try {
    } catch {
    } finally {
        let i = 1;
        do {
            if (i % 2 === 0) {
                return i;
                ~~~~~~~~~    [no-return-in-finally]
            }

            ++i;
        } while (i < 5);
    }
}

function() {
    try {
        if (foo()) {
            try {
                bar();
            } catch {
            } finally {
                return 2;
                ~~~~~~~~~    [no-return-in-finally]
            }
        }
    } catch {
    } finally {
        for (const i of [1, 2, 3, 4, 5]) {
            if (i % 2 === 0) {
                return 1;
                ~~~~~~~~~    [no-return-in-finally]
            }
        }
    }
}

function foo() {
    try {
        try {
        } finally {
            return 1;
            ~~~~~~~~~    [no-return-in-finally]
        }
    } catch {
    }
}

function foo() {
    try {
    } finally {
        throw "error";
        ~~~~~~~~~~~~~~    [no-throw-in-finally]
    }
}

function foo() {
    for (const i of [1, 2, 3, 4, 5]) {
        try {
        } finally {
            continue;
            ~~~~~~~~~    [no-continue-in-finally]
        }
    }
}

function foo() {
    for (const i of [1, 2, 3, 4, 5]) {
        try {
        } finally {
            break;
            ~~~~~~    [no-break-in-finally]
        }
    }
}

function foo() {
    for (let i = 0; i < 5; ++i) {
        try {
        } finally {
            continue;
            ~~~~~~~~~    [no-continue-in-finally]
        }
    }
}

function foo() {
    do {
        try {
        } finally {
            continue;
            ~~~~~~~~~    [no-continue-in-finally]
        }
    } while (i > 0);
}

function foo() {
    while (i > 0) {
        try {
        } finally {
            continue;
            ~~~~~~~~~    [no-continue-in-finally]
        }
    }
}

function foo() {
    do {
        try {
        } finally {
            break;
            ~~~~~~    [no-break-in-finally]
        }
    } while (i > 0);
}

function foo() {
    while (i > 0) {
        try {
        } finally {
            break;
            ~~~~~~    [no-break-in-finally]
        }
    }
}

function foo() {
    switch (a) {
        case 1:
            try {
                return;
            } finally {
                break;
                ~~~~~~    [no-break-in-finally]
            }
            break;
        default:
            try {
            } finally {
                break;
                ~~~~~~    [no-break-in-finally]
            }
            break;
    }
}

function foo() {
    label:
    try {
    } finally {
        break label;
        ~~~~~~~~~~~~    [no-break-in-finally]
    }
}

function foo() {
    label:
    for (let i = 0; i < 4; ++i) {
        try {
        } finally {
            break label;
            ~~~~~~~~~~~~    [no-break-in-finally]
        }
    }
}

function foo() {
    outer:
    for (let i = 0; i < 4; ++i) {
        try {
        } finally {
            inner:
            for (let j = 0; j < 2; ++j) {
                break outer;
                ~~~~~~~~~~~~    [no-break-in-finally]
            }
        }
    }
}

function foo() {
    label:
    for (let i = 0; i < 4; ++i) {
        try {
        } finally {
            continue label;
            ~~~~~~~~~~~~~~~    [no-continue-in-finally]
        }
    }
}

function foo() {
    outer:
    for (let i = 0; i < 4; ++i) {
        try {
        } finally {
            inner:
            for (let j = 0; j < 2; ++j) {
                continue outer;
                ~~~~~~~~~~~~~~~    [no-continue-in-finally]
            }
        }
    }
}

// valid
function() {
    try {
    } finally {
        (function() {
            return 1;
        })();
    }
}

// valid
function() {
    try {
        foo();
    } finally {
        class C {
            bar() {
                return 1;
            }
        }
    }

    return 1;
}

// valid
function() {
    try {
    } catch {
    } finally {
        for (let i = 0; i < 5; ++i) {
            if (i % 2 === 0) {
                continue;
            }
        }
    }
}

// valid
function() {
    try {
    } catch {
    } finally {
        for (let i = 0; i < 5; ++i) {
            if (i % 2 === 0) {
                break;
            }
        }
    }
}

// valid
function() {
    try {
    } catch {
    } finally {
        switch (a) {
            case 1:
                break;
            default:
                break;
        }
    }
}

//valid
function() {
    try {
    } finally {
        label:
            break label;
    }
}

//valid
function() {
    try {
    } finally {
        outer:
        for (let i = 0; i < 2; ++i) {
            if (i % 2 === 0) {
                continue outer;
            }

            inner:
            for (let j = 0; j < 2; ++j) {
                if (j % 2 === 0) {
                    continue outer;
                }
            }
        }
    }
}

[no-return-in-finally]: return statements in finally blocks are forbidden.
[no-throw-in-finally]: throw statements in finally blocks are forbidden.
[no-break-in-finally]: break statements in finally blocks are forbidden.
[no-continue-in-finally]: continue statements in finally blocks are forbidden.
